### PR TITLE
Background sync for datasets from CommCare HQ

### PR DESCRIPTION
**Pull request by kaapstorm**
Jira: [SC-3212](https://dimagi.atlassian.net/browse/SC-3212)

This pull request includes
* A series of small prefactor commits
* A large commit that adds an API endpoint for accepting dataset change requests from CommCare HQ
* A large commit that adds an OAuth 2.0 server for authenticating incoming requests
* A migration for adding the models for the OAuth 2.0 server

Easiest to review :blowfish: commit :octopus: by :tropical_fish: commit :fish: 

This code has been tested locally.

It will be deployed to Staging on Wed Feb 28.

QA is scheduled.

### CODE REVIEW

File: hq_superset/models.py
Comment (mkangia) [inline hq_superset/models.py:47]: is this deleting and then inserting the row for the same case instead of an update?
Not that I am opposed to the approach, just trying to understand the reasoning behind it.

File: hq_superset/api.py
Comment (mkangia) [inline hq_superset/api.py:46]: nit: do we want to keep a better url for this? I assume this is getting used internally and I don't fully understand where all this will be used. "webhook" seems quite open ended on what that could mean.

File: hq_superset/const.py
Comment (mkangia) [inline hq_superset/const.py:2]: nit: `CommCareHQ Data`?

**Review by mkangia — COMMENTED** [review]
thanks for the refactor commits @kaapstorm.

I tried reviewing this PR but it was a bit difficult with this PR attending to quite a lot of changes with some big commits.
Plus the part about OAuth that came along with syncing data sources is new to me.

Are you able to break this down further for easier review? Or else if its better the way it is, I will take a more dedicated pass. Please let me know

File: hq_superset/const.py
Comment (kaapstorm) [reply hq_superset/const.py:2]: Reply nit: "CommCare HQ" [has a space in it](https://www.commcarehq.org/). :joy: 

Unfortunately this value needs to match the database name in existing environments. Previously, users had to create this database manually (see [instructions in README.md](https://github.com/dimagi/hq_superset/blob/master/README.md#create-a-postgres-database-connection-for-storing-hq-data), specifically, "The name of the DISPLAY NAME should be 'HQ Data' exactly". So we're stuck with "HQ Data" (even though the convention in Superset is to use a lowercase word, like "commcare-hq-data").

File: hq_superset/models.py
Comment (kaapstorm) [reply hq_superset/models.py:47]: I wrangled with this for a long time. SQLAlchemy's [Postgres dialect supports upsert](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#insert-on-conflict-upsert). But there is a catch: The "doc_id" column must be indexed in order to use it as a constraint in the upsert's "where" clause.

That table is created by Pandas, and Pandas can't create an index on that column. You can create the table first, and then dump Pandas data into it, but I couldn't get Pandas to create the rest of columns and populate the table correctly. Maybe I was missing something. Eventually I decided to just delete and then insert.

Next time I have a few more hours to spend on it, I'll go back to it. Or we can make a follow-up ticket for nice-to-haves.

File: hq_superset/api.py
Comment (kaapstorm) [reply hq_superset/api.py:46]: Good point. "hq_datasource"? "hq_dataset"?

Comment (mkangia) [reply hq_superset/api.py:46]: how about just `dataset`? I see we have been following a convention to put a `hq_` in all names.
Is it to avoid conflict with superset routes?

Comment (mkangia) [reply hq_superset/api.py:46]: having `hq_` in all names makes me think that its coming from hq but hq does not have anything called dataset.
hence you also suggested `hq_datasource` but this view is changing dataset.

**Comment by kaapstorm** [comment]
Tests are failing because they are not configured correctly. I'll fix that.

@mkangia you asked:
> Are you able to break this down further for easier review? 

I agree that [Add OAuth 2.0 server](https://github.com/dimagi/hq_superset/pull/41/commits/dad8b91badb6c1d10d1f564f7bc585154a7ee842) is a monster. All the commits that went into it [can be found here](https://github.com/dimagi/hq_superset/pull/29) but I'm not sure that's any more helpful.

I'll try to unpack them. I'll timebox it and see how I get on. (Rearranging and squashing the commits from the draft PR took hours. Hopefully unpacking might be quicker.)

**Comment by mkangia** [comment]
Thanks @kaapstorm 

Good to timebox or else then If its easier, we could jump on a call where we can walk through the big commits

File: hq_superset/api.py
Comment (kaapstorm) [reply hq_superset/api.py:46]: > Is it to avoid conflict with superset routes?

Yes, exactly. Superset already has "dataset" routes.

> but this view is changing dataset.

Yes. ... Although it is the endpoint for DataSourceRepeaters. The URL for their ConnectionSettings will be 'hq_datasource/change". So I guess at least it makes sense in the context of the client/HQ.

Comment (mkangia) [reply hq_superset/api.py:46]: > Although it is the endpoint for DataSourceRepeaters

I am thinking that we should name things in here independent of what they are used for in CommCareHQ and more around what they mean here. I guess I will have a better suggestion of what this should be named if I look into this more deeply. Fine to leave it as it is then.
We are in the early phases of this product so would be nice to focus on good names early on.

File: hq_superset/api.py
Comment (Charl1996) [inline hq_superset/api.py:43]: @kaapstorm How was this limit determined?

File: hq_superset/api.py
Comment (Charl1996) [inline hq_superset/api.py:53]: @kaapstorm Quickly remind me again, is this endpoint going to be hit for every row change in a data source?

File: hq_superset/utils.py
Comment (Charl1996) [inline hq_superset/utils.py:214]: The docstring is amazing!

File: hq_superset/models.py
Comment (Charl1996) [inline hq_superset/models.py:50]: I'm going to be nitpicking here...

Can we maybe call this something like `HQOAuthClient`? The reason is that the current naming made me think this was the request client (don't know if I'm just super pedantic).

Anyway, I'll leave it to your own discretion...

File: hq_superset/services.py
Comment (Charl1996) [inline hq_superset/services.py:155]: Nit: can this be a separate function? (maybe something like `get_oauth_client_for_domain`)

File: hq_superset/services.py
Comment (Charl1996) [inline hq_superset/services.py:161]: Personally I'd put this in another function as well, just so `subscribe_to_hq_datasource` doens't have to know how the `webhook_url` is generated., but I'll leave it up to you...

File: README.md
Comment (Charl1996) [inline README.md:11]: Maybe more of an HQ discussion, but should we not have a different FF for this workflow so users don't have to be concerned about the `User Configurable Reports UI` feature flag until they actually want to user the UCRs?

**Comment by Charl1996** [comment]
Absolutely great work, @kaapstorm !

Just a couple of comments, nothing big.

File: hq_superset/models.py
Comment (mkangia) [reply hq_superset/models.py:47]: thanks @kaapstorm 
the current approach sounds good to me

**Comment by kaapstorm** [comment]
Please excuse the force-pushes. Ready for review @mkangia .

And thanks for the review @Charl1996 ! I'll follow up in the morning.

File: hq_superset/models.py
Comment (mkangia) [inline hq_superset/models.py:52]: nit: building on the conversation about naming, this could be `commcarehq_oauth_client` or `commcare_hq_oauth_client`

File: hq_superset/models.py
Comment (mkangia) [inline hq_superset/models.py:82]: Can we add a docstring to this function please?
I assume it simply revokes all current active tokens for the client.

And since I don't understand this syntax well, just confirming that we don't need `values(OAuth2Token.access_token_revoked_at=revoked_at)`?

File: hq_superset/services.py
Comment (mkangia) [inline hq_superset/services.py:161]: nit: should this be `hq_request.root_url`?
and same later in this function?

File: README.md
Comment (mkangia) [inline README.md:76]: Is this and `OAUTHLIB_INSECURE_TRANSPORT` needed if we are using OAuth with CommCareHQ environment available locally?
If we are providing this option to communicate with live CommCareHQ environments, we probably should not.

File: README.md
Comment (mkangia) [inline README.md:164]: Nice!

File: hq_superset/migrations/env.py
Comment (mkangia) [inline hq_superset/migrations/env.py:77]: Did we take this process from somewhere?
Just thinking why we would need offline/online processes.

File: README.md
Comment (mkangia) [inline README.md:159]: and then one should uncomment them back afterwards?

File: README.md
Comment (mkangia) [inline README.md:63]: did we rename `superset_meta` to `superset`?

File: README.md
Comment (mkangia) [inline README.md:94]: Just checking on why we removed this part?

File: hq_superset/migrations/README
Comment (mkangia) [inline hq_superset/migrations/README:1]: nit: we need this file?

File: hq_superset/migrations/alembic.ini
Comment (mkangia) [inline hq_superset/migrations/alembic.ini:63]: is it something that would need to be updated for production environments?

File: hq_superset/tests/config_for_tests.py
Comment (mkangia) [inline hq_superset/tests/config_for_tests.py:33]: nit: May be we need to update README to add note to add this DB to run tests.

File: hq_superset/utils.py
Comment (mkangia) [inline hq_superset/utils.py:269]: nit: probably its just me, though when I see `row`, I assume its a list. But here, this is a dictionary. That got me.

**Review by mkangia — COMMENTED** [review]
Hey @kaapstorm 

Thanks for the work on this. This is a tricky one.
Nice refactor and clean up to make the repo better in general.

It was a big one so a lot of comments, mostly minor.

Just one critical query about why we chose to use OAuth instead of token based authentication to get updates from CommCareHQ.

File: hq_superset/services.py
Comment (kaapstorm) [reply hq_superset/services.py]: ba6286d

File: hq_superset/services.py
Comment (kaapstorm) [reply hq_superset/services.py]: 9f4b053

File: hq_superset/models.py
Comment (kaapstorm) [reply hq_superset/models.py]: d30cc79

File: hq_superset/api.py
Comment (kaapstorm) [reply hq_superset/api.py:53]: This endpoint will be hit for every change for a doc_id, where a doc_id originally refers to a case or a form. Sometimes one doc_id can result in multiple rows, e.g. a data source definition that pulls out a question inside a repeat group.

File: hq_superset/models.py
Comment (kaapstorm) [reply hq_superset/models.py]: When I rebased (before reading this comment) I renamed this class "OAuth2Client", in line with its mixin. I did the same with "Token".

"HQOAuthClient" may have been a better name, considering it's got a "domain" field, but I feel like "OAuth2Client" is close enough, and better than "CommCareHQOAuth2Client".

File: hq_superset/utils.py
Comment (kaapstorm) [reply hq_superset/utils.py:271]: That's what the type hint on line 256 is for.

While "row" could be a list, or a tuple, or a dictionary, "data" could be absolutely anything! Any variable contains data, doesn't it? And "table" is also ambiguous -- SQLAlchemy documentation uses the variable name "table" to refer to two or three different types.

So type hints, _in this specific case_, are helpful. The alternative would be to rename "data" to something more accurate, but less friendly, like "list_of_column_value_pairs", but then you wouldn't get any validation from your IDE.

Fun exercise: Open this in your IDE. Ctrl-click on the "columns" property of "table". Magic, right? Now delete the type hint, and try again.

Comment (kaapstorm) [reply hq_superset/utils.py:271]: (You kinda handed me my favorite drum, and I just went ahead and banged it. :joy: :speak_no_evil:  Sorry about that.)

File: hq_superset/tests/config_for_tests.py
Comment (kaapstorm) [reply hq_superset/tests/config_for_tests.py:33]: You don't need to add it to run tests. `get_hq_database()` creates it automatically if it is missing.

File: hq_superset/migrations/alembic.ini
Comment (kaapstorm) [reply hq_superset/migrations/alembic.ini:63]: Production environments use `superset db upgrade` to run migrations, and that should use [env.py to set it automatically](https://github.com/dimagi/commcare-analytics/blob/875ff55342997eb94e7dfc76fb3a4ebcaa862129/hq_superset/migrations/env.py#L17) for that environment.

File: hq_superset/migrations/README
Comment (kaapstorm) [reply hq_superset/migrations/README]: It was generated by Alembic. We don't really need it. 0974df5

File: README.md
Comment (kaapstorm) [reply README.md:94]: This [gets done automatically](https://github.com/dimagi/commcare-analytics/blob/d30cc792ecd236157356504bda5270fd9430a461/hq_superset/utils.py#L47) now by `get_hq_database()`.

File: hq_superset/api.py
Comment (kaapstorm) [reply hq_superset/api.py]: f0276dd

File: hq_superset/models.py
Comment (kaapstorm) [reply hq_superset/models.py:41]: We are addressing deletions. When all the rows for a form or a case are deleted, then `data` will be an empty list.

File: hq_superset/models.py
Comment (kaapstorm) [reply hq_superset/models.py:29]: Yes. In fact, the table belongs to a schema which includes the domain name. That happens [when the table is created here](https://github.com/dimagi/commcare-analytics/blob/ba6286d124b483fe03eb527e0971cfa59e8502b0/hq_superset/services.py#L90). So the full name of the table for data source "abc123" in domain "foo" will be "hqdomain_foo.abc123".

File: hq_superset/hq_domain.py
Comment (kaapstorm) [reply hq_superset/hq_domain.py]: aedcb42

**Comment by kaapstorm** [comment]
> one critical query about why we chose to use OAuth instead of token based authentication

OAuth 2.0 is token-based authentication. ([More info here](https://www.oauth.com/oauth2-servers/access-tokens/).) CommCare HQ already supports OAuth 2.0 token-based authentication for repeaters.

Do you have a different kind of token-based authentication in mind? What standard are you thinking of?

File: hq_superset/models.py
Comment (Charl1996) [reply hq_superset/models.py]: > I feel like "OAuth2Client" is close enough

Perfect!

File: README.md
Comment (mkangia) [reply README.md:94]: Oh, isn't this part about changing the auth_type to enable login via admin which is still applicable?

**Comment by mkangia** [comment]
> Do you have a different kind of token-based authentication in mind? What standard are you thinking of?

Resolved offline.
OAuth is the better choice keeping security in mind and the library is already available via superset/flask so no new dependency was added.

File: README.md
Comment (kaapstorm) [reply README.md:94]: Good point! We don't need to log in as admin to create the database any more, but it is useful to know how to log in as admin if there is any other reason to do so. I'll add that back.

File: hq_superset/utils.py
Comment (mkangia) [reply hq_superset/utils.py:271]: I kinda new what I was getting back when i asked that question, I knew you would find your way to typehints.

File: README.md
Comment (kaapstorm) [reply README.md:94]: af4878f

File: README.md
Comment (mkangia) [inline README.md:114]: nit: this can't be done in a production environment right? This is only for local setup?

File: hq_superset/api.py
Comment (mkangia) [reply hq_superset/api.py:18]: this was taken offline.

File: hq_superset/models.py
Comment (mkangia) [reply hq_superset/models.py:42]: thanks, can we document that somewhere in this function?

**Comment by mkangia** [comment]
New commits look good. Will wait for responses on other comments from the last review.

File: README.md
Comment (kaapstorm) [reply README.md:114]: :thinking: Doing this in a production environment will prevent CommCare HQ users from logging in. But for self-hosted environments it could be useful for maintenance or setting up additional databases, etc.

Do you think I should add more context? Or maybe give advice on when to use CommCare Ansible, and the rare cases when it might be appropriate to edit the config directly?

Comment (mkangia) [reply README.md:114]: >  Doing this in a production environment will prevent CommCare HQ users from logging in

I believe that is true hence this should not be used for production environments.

I was very confused when I saw this for the first time. For me, I have always thought of this as something to be used only locally.
Though now when you mention maintenance, I see there is no way to access UI for an admin without this change, so it is the only way out.

> the rare cases when it might be appropriate to edit the config directly

Right, we should just caution that doing this change disables usual new user login (I assume the ones who are logged in will continue to work), so, on a production environment this should be only done when system isn't in use.

P.S: why do we need this change though? I comment about that would be super helpful in the code

File: hq_superset/models.py
Comment (kaapstorm) [reply hq_superset/models.py:49]: 2c0c766

File: README.md
Comment (kaapstorm) [reply README.md:118]: ceb8f80

**Comment by mkangia** [comment]
New commits look good :+1:

**Comment by mkangia** [comment]
Hey @kaapstorm 

I see some open comments from my earlier review. Waiting on those to do a full pass again. I hope these will be quick since they are mostly minor.

**Review by Charl1996 — APPROVED** [review]
Did a brief review of the newer commits and all seems fine. I'm happy with this PR.

File: hq_superset/services.py
Comment (Charl1996) [inline hq_superset/services.py:162]: (related to [this](https://github.com/dimagi/commcare-hq/pull/34179#issuecomment-1985185551) comment)

[This](https://stackoverflow.com/questions/66663507/flask-url-for-http-instead-of-https) is an interesting post. It seems like one might need to explicitly specify the `https` scheme should be used.

File: hq_superset/oauth2_server.py
Comment (mkangia) [inline hq_superset/oauth2_server.py:45]: :spiral_notepad:

**Comment by kaapstorm** [comment]
49d575aa47411ba211e1dda3ec260600c6dd64c4 fixed the problem with syncing from CommCare Staging. I don't like it, but the [right way to do it](https://stackoverflow.com/a/45333882) didn't work.

"If at first (or second, or third, or fourth) you don't succeed, use a bigger hammer."

File: hq_superset/services.py
Comment (kaapstorm) [reply hq_superset/services.py]: 9042fb6e37d1e5a2f9d9a19e264f62c615ce09fd

File: README.md
Comment (kaapstorm) [reply README.md:11]: I think we can definitely pull CommCare Analytics into the current discussion around UCR feature flags.

**Comment by kaapstorm** [comment]
I think that's everything. If there is anything I've missed, please link to it.

File: hq_superset/__init__.py
Comment (mkangia) [inline hq_superset/__init__.py:30]: Hey @kaapstorm 

Good find on the proxy redirection.

I can make peace with this change to unblock QA/UAT.

It's a positive that this is on analytics server that we have control/access to and not superset.

I was trying to find where in nginx conf, this is getting set and compare that with what we have for HQ though HQ nginx template has a lot happening.

Couple of things I noticed that could be faulty
1. we have two "server" blocks with the same server name in nginx conf. I am not sure if that is expected/correct
2. I don't know what the following block does
```
if (!-f $request_filename) {
      proxy_pass http://superset;
      break;
    }
```

@sravfeyn assuming you have more insight when this was setup.
Can you see something faulty with the current nginx conf or point out what could redirecting urls to http?
I don't know what this block does

Comment (kaapstorm) [reply hq_superset/__init__.py:30]: I took a quick look at the nginx config, and realised it will take me a bit longer to change and test it. I'll merge this PR, and I'll create a follow up ticket to see how we can configure nginx so that we can use werkzeug's "ProxyFix" class (or not have to use it at all) instead of using this environment variable.

Comment (mkangia) [reply hq_superset/__init__.py:30]: Okay @kaapstorm, thanks for taking a look.

I am fine with this merge till this stays on staging and is not released to production environment until the secure fix is done.

**Pull request merged**
